text/3626-format-args-implicit-dot.md

```diff
@@ -131,9 +131,9 @@ format_args!("{unique_identifier}", unique_identifier=name.field1.field2)
 ```
 
 Any `Deref` operations or `.await` operations associated with the `.` in each
-format argument are evaluated from left-to-right as they appear in the format
-string, at the point where the format string argument is evaluated, before the
-positional or named arguments are evaluated.
+format argument are evaluated exactly once, from left-to-right as they appear
+in the format string, at the point where the format string argument is
+evaluated, before the positional or named arguments are evaluated.
```

Comment:
yes. in particular currently `format_args!("{x} {x}")` is lowered to `format_args!("{0} {0}", x)`.

the RFC should specify that `format_args!("{x.y} {x.y}")` should lower to `format_args!("{0} {1}", x.y, x.y)` rather than `format_args!("{0} {0}", x.y)`.

Reply:
deduplication should only be allowed when the compiler can show that the effect is identical to the user code being run for each duplicate, so it is allowed to deduplicate if there is no deref or if the deref is idempotent.

Reply:
i don't think the compiler should be able to deduce whether `x.y` has side effect or not *during macro expansion*.

is there any disadvantage in never de-duplicating?

Reply:
well, `format_args!` actually [expands to its own AST item](https://github.com/rust-lang/compiler-team/issues/541) rather than expanding to Rust syntax like normal macros, it seems reasonable that there will eventually be a corresponding `FormatArgs` item when rustc does type checking, rustc can deduplicate once it knows argument types since it can easily see if any `deref` calls were made and if those `deref` calls are to known-idempotent `deref` implementations (such as `LazyCell`, `Arc`, `Box`, etc.). eventually it might look at the `deref` function's body to see if it's idempotent.

Reply:
@programmerjake wrote:
> I think you need to specify that duplicate format arguments are evaluated N times when there are N duplicates -- aka. deduplication does not occur.

Done.